### PR TITLE
Invite system

### PR DESCRIPTION
**Pull request by magyarosibotond**
Hey @anthonynsimon,

I've created this WIP pull request to discuss the technical decision for the pull request while moving forward with the implementation.

I've created a new controller for /projects/:id/invites (since it represents a separate feature of the app) and I was thinking about where to move the invite / add to project logic.

The webapp shouldn't know about the user when it tries to invite someone to the project.

I was thinking about removing POST /project/:id/users and move the logic to POST /projects/:id/invites.

What do you think?

### CODE REVIEW

**Comment by anthonynsimon** [comment]
@magyarosibotond hey, I think it's a great start. I agree that it makes sense to then move the "add project user" logic into this new controller.

Regarding the rest of the implementation, what would you say about something like the following:

An admin invites a user to the project, two possible cases:
- Invited user is already registered on traduora with that email, then simply auto-assign as team member and skip the invite, just send email notification to user.
- Invited user is not yet registered on traduora, then send invite email and assign as team member on account creation.

Basically we would require to check the invites table when a user is signing up, to create the ProjectUser entitiy if there are any open invites.

**Comment by magyarosibotond** [comment]
Hey @anthonynsimon! 

I've just managed to squeeze some contribution into my timeframe.

Here's what's done:
- CRUD operations for invites.
- Email notification for invite (URL needs to be changed).
- Show invites (email, role) under the Team tab.

Here's what needs to be done:

- Use the invite code at sign up. **I was wondering to add the invite code as query parameter and send it in the request body for sign up. What do you think?** 
- Add foreign key the invite table. **How do you generate the random keys? Can I call them on a more descriptive name?** 
- Tests and code clean up.

If everything goes well, I might be able to finish this feature by the end of the week.

Could you take a look at the current progress?

Thanks!

**Comment by magyarosibotond** [comment]
Hey @anthonynsimon!

I'm done with the changes. Tests still needed. 

Could you take a look?

Thanks!

File: api/src/controllers/auth.controller.ts [hunk 1]
```diff
@@ -38,18 +40,45 @@ export class AuthController {
...
     @InjectRepository(ProjectClient) private projectClientRepo: Repository<ProjectClient>,
   ) {}
 
   @Post('signup')
   @HttpCode(HttpStatus.OK)
   async signup(@Body() payload: SignupRequest) {
     if (!config.signupsEnabled) {
-      throw new ForbiddenException('Signups are disabled');
+      let invitesCount = await this.inviteRepo.count({
+        where: { email: payload.email, status: InviteStatus.Sent },
```
Comment (anthonynsimon) [inline api/src/controllers/auth.controller.ts]: You gotta normalize the incoming email to make sure it'll look the same as the one in the DB.

File: api/src/controllers/auth.controller.ts [hunk 2]
```diff
@@ -38,18 +40,45 @@ export class AuthController {
...
     const user = await this.userService.create(payload.name, payload.email, payload.password);
 
+    // accept project invites
+
+    const invites = await this.inviteRepo.find({
+      where: { email: payload.email, status: InviteStatus.Sent },
+      relations: ['project'],
+    });
+    invites.forEach(invite => (invite.status = InviteStatus.Accepted));
+    await this.inviteRepo.save(invites);
```
Comment (anthonynsimon) [inline api/src/controllers/auth.controller.ts]: Since there's many things being saved simultaneously, it'd be good to wrap them in a transaction in case something goes wrong in the middle of it. For reference see https://github.com/traduora/traduora/blob/2882ecc9c9a8bfae5b7279448ffa261b7e0e0b4c/api/src/controllers/term.controller.ts#L46

File: webapp/src/app/projects/components/team-overview/team-overview.component.ts [hunk 3]
```diff
@@ -18,7 +20,7 @@ export class TeamOverviewComponent implements OnInit, OnDestroy {
   @Select(ProjectsState.currentProject)
   project$: Observable<Project | undefined>;
 
-  @Select(ProjectUsersState.isLoading)
+  @Select(ProjectUsersState.isLoading && ProjectInviteState.isLoading)
```
Comment (anthonynsimon) [inline webapp/src/app/projects/components/team-overview/team-overview.component.ts]: Does this actually work? I think you might have to make two individual selects and combine the observables manually.

File: webapp/src/app/projects/stores/project-invite.state.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+import { catchError, finalize, tap } from 'rxjs/operators';
+import { Logout } from '../../auth/stores/auth.state';
+import { errorToMessage } from '../../shared/util/api-error';
+import { ProjectRole } from '../models/project-role';
+import { ClearCurrentProject, SetCurrentProject } from './projects.state';
+import { ProjectInvite } from '../models/project-invite';
+import { ProjectInviteService } from '../services/project-invite.service';
+
+export class ClearMessages {
+  static readonly type = '[ProjectUser] Clear messages';
```
Comment (anthonynsimon) [inline webapp/src/app/projects/stores/project-invite.state.ts]: It should be [ProjectInvite] in the message :)

File: webapp/src/app/projects/stores/project-invite.state.ts [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+  @Selector()
+  static invites(state: ProjectInviteStateModel) {
+    return state.invites;
+  }
+
+  ngxsOnInit(ctx: StateContext<ProjectInviteStateModel>) {}
+
+  @Action(SetCurrentProject)
+  setCurrentProject(ctx: StateContext<ProjectInviteStateModel>, action: SetCurrentProject) {
```
Comment (anthonynsimon) [inline webapp/src/app/projects/stores/project-invite.state.ts]: I think you can remove the hook on the setCurrentProject, the invites can be loaded only when the component needs it.

File: api/src/controllers/project-invite.controller.ts [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+    return {
+      data: invites,
+    };
+  }
+
+  @Post(':projectId/invites')
+  async create(@Req() req, @Param('projectId') projectId: string, @Body() inviteUserRequest: InviteUserRequest) {
+    const user = this.auth.getRequestUserOrClient(req, { mustBeUser: true });
+    await this.auth.authorizeProjectAction(user, projectId, ProjectAction.AddProjectUser);
```
Comment (anthonynsimon) [inline api/src/controllers/project-invite.controller.ts]: We should probably rename this ProjectAction to `InviteProjectUser` for consistency and Remove the `AddProjectUser` action since it's not supported other than via invites. Just be careful to refactor all the existing usages also in the frontend.

File: api/src/controllers/project-invite.controller.ts [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    private mail: MailService,
+    @InjectRepository(User) private userRepo: Repository<User>,
+    @InjectRepository(Invite) private inviteRepo: Repository<Invite>,
+    @InjectRepository(ProjectUser) private projectUserRepo: Repository<ProjectUser>,
+  ) {}
+
+  @Get(':projectId/invites')
+  async find(@Req() req, @Param('projectId') projectId: string) {
+    const user = this.auth.getRequestUserOrClient(req, { mustBeUser: true });
+    await this.auth.authorizeProjectAction(user, projectId, ProjectAction.ViewProjectUsers);
```
Comment (anthonynsimon) [inline api/src/controllers/project-invite.controller.ts]: We can add a new ProjectAction called ViewProjectInvites and authorize only admins, so that editors and viewers do not see pending invites.

File: webapp/src/app/projects/components/team-overview/team-overview.component.html [hunk 8]
```diff
@@ -14,5 +14,14 @@ <h5 class="font-serif m-0" i18n>Project Team</h5>
         (remove)="removeProjectUser(project.id, $event.userId)"
       ></app-team-member>
     </div>
+    <div *ngFor="let invite of (projectInvites$ | async)">
```
Comment (anthonynsimon) [inline webapp/src/app/projects/components/team-overview/team-overview.component.html]: Maybe we can add a little subtitle or something for hinting to the user that what follows are the invites.

File: api/src/controllers/auth.controller.ts (hunk 1 above)
Comment (magyarosibotond) [reply api/src/controllers/auth.controller.ts]: Moved the validation to a different file.

File: api/src/controllers/auth.controller.ts (hunk 2 above)
Comment (magyarosibotond) [reply api/src/controllers/auth.controller.ts]: Is there a way to make transations in multiple repositories at the same time?

File: webapp/src/app/projects/components/team-overview/team-overview.component.ts (hunk 3 above)
Comment (magyarosibotond) [reply webapp/src/app/projects/components/team-overview/team-overview.component.ts]: done

File: webapp/src/app/projects/stores/project-invite.state.ts (hunk 4 above)
Comment (magyarosibotond) [reply webapp/src/app/projects/stores/project-invite.state.ts]: done

File: webapp/src/app/projects/components/team-overview/team-overview.component.html (hunk 8 above)
Comment (magyarosibotond) [reply webapp/src/app/projects/components/team-overview/team-overview.component.html]: This is what I came up with:

Screenshot 2019-04-02 at 17 45 32

File: api/src/controllers/auth.controller.ts (hunk 2 above)
Comment (anthonynsimon) [reply api/src/controllers/auth.controller.ts]: yes, typeorm can figure out the repos from the entities when using a transaction manager, see for example: https://github.com/traduora/traduora/blob/master/api/src/controllers/translation.controller.ts#L63

File: webapp/src/app/projects/components/team-overview/team-overview.component.html (hunk 8 above)
Comment (anthonynsimon) [reply webapp/src/app/projects/components/team-overview/team-overview.component.html]: Looks good!

File: webapp/src/app/projects/stores/project-invite.state.ts (hunk 5 above)
Comment (anthonynsimon) [reply webapp/src/app/projects/stores/project-invite.state.ts]: Here's were you can dispatch the get invites: https://github.com/traduora/traduora/blob/master/webapp/src/app/projects/components/team-overview/team-overview.component.ts#L37

Comment (anthonynsimon) [reply webapp/src/app/projects/stores/project-invite.state.ts]: And I think you might still need to add the functions for the update invite and remove invite on the team overview component :)

File: api/src/controllers/project-invite.controller.ts (hunk 6 above)
Comment (magyarosibotond) [reply api/src/controllers/project-invite.controller.ts]: makes sense

File: api/src/controllers/auth.controller.ts (hunk 2 above)
Comment (magyarosibotond) [reply api/src/controllers/auth.controller.ts]: thanks! done.

File: webapp/src/app/projects/stores/project-invite.state.ts (hunk 5 above)
Comment (magyarosibotond) [reply webapp/src/app/projects/stores/project-invite.state.ts]: that should be added.

**Comment by anthonynsimon** [comment]
Hi @magyarosibotond , have you had time to look into the remaining points to be able to merge it? It'd be great to have this new feature in the next release :)

**Comment by magyarosibotond** [comment]
Welcome back @anthonynsimon! I did manage to resolve the resolve the remaining points, and I just started writing tests for the added feature.

**Comment by anymaniax** [comment]
Hello, I did a pull request on your branch to fix your tests

**Comment by anthonynsimon** [comment]
Closing as #66 now hosts the remaining changes.
